ext/NNlibEnzymeCoreExt/NNlibEnzymeCoreExt.jl

```diff
@@ -0,0 +1,374 @@
+    else
+        nothing
+    end
+    shadow = if EnzymeCore.EnzymeRules.needs_shadow(config)
+        y.dval
+    else
+        nothing
+    end
+
+    # Cache x if its overwritten and w is active (and thus required)
+    cache_x = ( EnzymeCore.EnzymeRules.overwritten(config)[3]
+                && !(typeof(w) <: EnzymeCore.Const)
+                && !(typeof(y) <: EnzymeCore.Const)
+                ) ? copy(x.val) : nothing
+
+    # Cache w if its overwritten and x is active (and thus required)
+    cache_w = ( EnzymeCore.EnzymeRules.overwritten(config)[4]
+                && !(typeof(x) <: EnzymeCore.Const)
+                && !(typeof(y) <: EnzymeCore.Const)
+                ) ? copy(w.val) : nothing
+
+    cache = (cache_x, cache_w)
+
+    return EnzymeCore.EnzymeRules.AugmentedReturn(primal, shadow, cache)
+end
+
+function EnzymeCore.EnzymeRules.reverse(config, func::EnzymeCore.Const{$name}, ::Type{RT}, cache,
+                                                y::EnzymeCore.Annotation{<:AbstractArray{yT, N}},
+                                                x::EnzymeCore.Annotation{<:AbstractArray{xT, N}},
+                                                w::EnzymeCore.Annotation{<:AbstractArray{wT, N}},
+                                                cdims; kwargs...) where {RT, yT, xT, wT, N}
+    cache_x, cache_w = cache
+
+    # Don't cache x if not overwritten and w is active (and thus required)
+    if !(typeof(w) <: EnzymeCore.Const) && !(typeof(y) <: EnzymeCore.Const)
+        if !EnzymeCore.EnzymeRules.overwritten(config)[3]
+            cache_x = x.val
+        end
+    end
+
+    # Don't cache w if not overwritten and x is active (and thus required)
+    if !(typeof(x) <: EnzymeCore.Const) && !(typeof(y) <: EnzymeCore.Const)
+        if !EnzymeCore.EnzymeRules.overwritten(config)[4]
+            cache_w = w.val
+        end
+    end
+
+    dys = y.dval
+    dxs = (typeof(x) <: EnzymeCore.Const) ? dys : x.dval
+    dws = (typeof(w) <: EnzymeCore.Const) ? dys : w.dval
+
+    if EnzymeCore.EnzymeRules.width(config) == 1
+        dys = (dys,)
+        dxs = (dxs,)
+        dws = (dws,)
+    end
+
+    for (dy, dx, dw) in zip(dys, dxs, dws)
+        if !(typeof(y) <: EnzymeCore.Const) && dy !== y.val
+
+            if !(typeof(x) <: EnzymeCore.Const) && dx !== x.val
+                # dx += grad wrt x.val
+                $dataname(dx, dy, cache_w, cdims.val; alpha=xT(1), beta=xT(1), kwargs...)
+            end
+            if !(typeof(w) <: EnzymeCore.Const) && dw !== w.val
+                # dw += grad wrt w.val
+                $filtername(dw, cache_x, dy, cdims.val; alpha=wT(1), beta=wT(1), kwargs...)
+            end
+            
+            dy .= 0
+        end
+    end
+
+    return (nothing, nothing, nothing, nothing)
+end
+
+end
+end
+
+function EnzymeCore.EnzymeRules.augmented_primal(config, func::EnzymeCore.Const{typeof(NNlib.gather!)}, ::Type{RT}, dst::OutType, src, idx::EnzymeCore.Const) where {OutType, RT}
```

Comment:
Are you basing these comments off some JuliaFormatter config? Because we don't have one for NNlib at present. We should have one, but I didn't want to hold up any PRs just because we don't.

Reply:
no just generic style comments